File: core/src/invariants.ts
/**
 * Invariants — K's shared, executable vocabulary of "what must always be
 * true", exported as a library rather than buried in K's own tests.
 *
 * One definition, three consumers:
 *   1. K's harness teeth (assert after scripted steps)
 *   2. the deterministic simulator (assert after EVERY effect, every seed)
 *   3. adopters — in their own tests and simulations, alongside their own
 *      app invariants registered in the same shape
 *
 * Why this shape: an invariant is a pure predicate over an observable
 * snapshot. It never reaches into internals, so the same check works
 * in-process, in simulation, and black-box against `status --json`.
 * Returning a *reason string* (not a boolean) means a violation explains
 * itself wherever it fires — including from a seed replay hours later.
 *
 * ---------------------------------------------------------------------------
 * SCOPE — what "the service" means here, stated so nothing is assumed:
 *
 *   K reasons about exactly ONE service identity per Upgrader: the process(es)
 *   its HostAdapter starts and stops. It is NOT a process supervisor for the
 *   machine and knows nothing about other processes that happen to run the
 *   same binary.
 *
 *   - swap profile: concurrent old-version processes are NORMAL (think many
 *     open CLI sessions). They are not K-managed, so no invariant here says
 *     anything about them. `never-dual-run` is vacuous in this profile rather
 *     than accidentally satisfied.
 *   - service / hosted: exactly one live incarnation of that identity, which
 *     is what `never-dual-run` constrains.
 *
 *   PROCESS MODEL (v0, declared rather than inferable from call order):
 *     EXCLUSIVE HANDOFF. At most one incarnation of the K-managed identity is
 *     alive at a time; a handoff is quiesce -> stop(old) -> start(new) -> prove.
 *     There is a downtime window by construction, and K makes it short rather
 *     than pretending it does not exist.
 *
 *     Zero-downtime overlap (start new, drain old, then stop old) is a
 *     DIFFERENT process model, not a flag on this one: it has a legitimate
 *     "both alive" state, needs a drain contract instead of quiesce, and
 *     changes which invariants hold. It is NOT implemented (ruled niche for
 *     v0). Anything claiming K supports it would be claiming a guarantee K
 *     does not make.
 *
 *   OUT OF SCOPE for v0 (deliberately, not by oversight): one binary running
 *   several instances with different arguments. One Upgrader owns one service
 *   identity; run one Upgrader per instance (separate stateDir), or have the
 *   adapter treat the whole set as a single identity. K does not orchestrate
 *   across instances, and no invariant here should be read as if it did.
 * ---------------------------------------------------------------------------
 */
import type { TxnPhase } from "./txn/state.ts";

/** What an invariant is allowed to see. Mirrors `status --json`. */
export interface WorldSnapshot {
  phase: TxnPhase;
  slots: { stable: string | null; experiment: string | null };
  /** Live service incarnations observed right now. */
  liveProcesses: Array<{ slot: "stable" | "experiment"; pid: number; startId: string; version: string }>;
  /** Journal intents in append order. */
  journalIntents: TxnPhase[];
  /**
   * Opaque, comparable digest of host workload state (sessions/jobs).
   * Hosts that preserve workloads across upgrade emit a stable digest.
   */
  workloadDigest?: string;
  /**
   * startId of the incarnation that was live BEFORE the current transition.
   * Present in real status readbacks; lets oracles prove a restart actually
   * happened rather than trusting a version string (Raft #5245 family).
   */
  priorIncarnationStartId?: string;
  /** Who owns this install: ourselves, or an external manager (PM/injector)? */
  installOwnership?: "self" | "managed-elsewhere";
}

/** null = holds; string = why it was violated (shown to humans and in seed replays). */
export type InvariantResult = string | null;

/**
 * What the HOST must do for a conditional guarantee to hold (assume-guarantee).
 * The left side of "if the app satisfies A, K guarantees G" — the app's
 * responsibility, verifiable by the harness against its adapter. Closed set:
 * if you need a new one, it is a design decision, not a free-text note.
 */
export type HostAssumption =
  | "quiesce-resume-inverse"        // parked workloads are unchanged while parked
  | "probe-from-live-process"       // evidence comes from the running process, not files
  | "compatibility-declared"        // checkCompatibility is implemented and correct
  | "data-format-backward-compatible" // version N can read what N+1 wrote
  | "exclusive-handoff"             // one live incarnation at a time: stop old, then start new
  | "resident-service";             // the host runs a process that is supposed to BE alive

export interface Invariant<S = WorldSnapshot> {
  id: string;
  description: string;
  /**
   * Empty/absent = K guarantees this UNCONDITIONALLY.
   * Non-empty = K guarantees it only if the host satisfies these assumptions.
   * This is how an adopter tells, from the library alone, which guarantees are
   * ours and which depend on their own code.
   */
  assumes?: readonly HostAssumption[];
  check: (snapshot: S) => InvariantResult;
}

/**
 * Never two live incarnations OF THE SERVICE IDENTITY K MANAGES.
 * Says nothing about other processes running the same binary (see SCOPE):
 * in the swap profile liveProcesses is empty, so this is vacuous by design.
 */
export const neverDualRun: Invariant = {
  id: "k.never-dual-run",
  description:
    "at most one live incarnation of the K-managed service identity (vacuous in the swap profile)",
  // NOT a universal truth about upgrades — it holds under K's declared process
  // model (exclusive handoff). An app that starts the new incarnation before
  // draining the old one is not violating a law of nature; it is running a
  // model K does not implement (see PROCESS MODEL below).
  assumes: ["exclusive-handoff"],
  check: (s) =>
    s.liveProcesses.length > 1
      ? `${s.liveProcesses.length} live incarnations: ${s.liveProcesses
          .map((p) => `${p.slot}#${p.pid}@${p.version}`)
          .join(", ")}`
      : null,
};

/**
 * Never a host with nothing runnable. Concerns the SLOTS (bytes on disk), not
 * whether a process is currently up: a stopped service with intact stable
 * bytes is not bricked.
 */
export const neverBricked: Invariant = {
  id: "k.never-bricked",
  description: "at least one slot always holds runnable bytes (about slots, not liveness)",
  check: (s) =>
    s.slots.stable === null && s.slots.experiment === null
      ? "both slots empty — nothing left to run"
      : null,
};

/**
 * A live incarnation's version matches the slot it was started from.
 * Only constrains processes K started; slots reporting null are ignored.
 */
export const liveProcessMatchesSlot: Invariant = {
  id: "k.live-process-matches-slot",
  description: "a live incarnation's version matches the slot it was started from",
  check: (s) => {
    for (const p of s.liveProcesses) {
      const slotVersion = s.slots[p.slot];
      if (slotVersion !== null && slotVersion !== p.version) {
        return `${p.slot} slot holds ${slotVersion} but a live process reports ${p.version}`;
      }
    }
    return null;
  },
};

/** Intent is journaled before the phase is entered (WAL ordering). */
export const journalPrecedesPhase: Invariant = {
  id: "k.journal-precedes-phase",
  description: "the current phase was journaled before being entered",
  check: (s) =>
    s.phase !== "idle" && !s.journalIntents.includes(s.phase)
      ? `phase ${s.phase} entered with no journal entry for it`
      : null,
};

/**
 * After promote/rollback the experiment slot is cleared. This is about the
 * TRANSACTION's own bookkeeping, not about whether a restart has happened yet.
 */
export const terminalLeavesNoExperiment: Invariant = {
  id: "k.terminal-leaves-no-experiment",
  description: "after promote/rollback the experiment slot is cleared",
  check: (s) =>
    (s.phase === "promoted" || s.phase === "rolled-back") && s.slots.experiment !== null
      ? `phase ${s.phase} but experiment slot still holds ${s.slots.experiment}`
      : null,
};

/** A copy owned by another manager never upgrades itself. */
export const managedCopyNeverSelfUpgrades: Invariant = {
  id: "k.managed-copy-never-self-upgrades",
  description: "an install owned by an external manager performs no upgrade of its own",
  check: (s) =>
    s.installOwnership === "managed-elsewhere" && s.phase !== "idle"
      ? `install is managed elsewhere but a transaction reached phase ${s.phase}`
      : null,
};

/** The built-in set. Adopters concat their own app invariants. */
export const BUILT_IN_INVARIANTS: readonly Invariant[] = [
  neverDualRun,
  neverBricked,
  liveProcessMatchesSlot,
  journalPrecedesPhase,
  terminalLeavesNoExperiment,
  managedCopyNeverSelfUpgrades,
];

/**
 * Deliberately NOT snapshot invariants — stated here so their absence is a
 * decision, not an oversight:
 *  - "no side effect before consent" and "every reconcile is journaled" are
 *    properties of an EVENT SEQUENCE, not of a single state. Forcing them into
 *    a snapshot predicate would require smuggling history into the snapshot;
 *    they are teeth over a scenario trace instead (policy/provenance teeth).
 */

export interface Violation {
  invariantId: string;
  reason: string;
}

/** Check a snapshot against a set; returns every violation (not just the first). */
export function checkInvariants(
  snapshot: WorldSnapshot,
  invariants: readonly Invariant[] = BUILT_IN_INVARIANTS,
): Violation[] {
  const out: Violation[] = [];
  for (const inv of invariants) {
    const reason = inv.check(snapshot);
    if (reason !== null) out.push({ invariantId: inv.id, reason });
  }
  return out;
}

/**
 * Workload preservation across a transition. Kept separate because it
 * compares two snapshots rather than judging one — hosts that declare a
 * workloadDigest get this for free, in upgrade AND rollback paths.
 */
export const WORKLOAD_PRESERVED_ASSUMES: readonly HostAssumption[] = ["quiesce-resume-inverse"];

export function workloadPreserved(before: WorldSnapshot, after: WorldSnapshot): InvariantResult {
  if (before.workloadDigest === undefined || after.workloadDigest === undefined) return null;
  return before.workloadDigest === after.workloadDigest
    ? null
    : `workload digest changed across the transition: ${before.workloadDigest} -> ${after.workloadDigest}`;
}

File: src/invariants.rs
//! Shared observation-based checks for adopters, fault harnesses and simulations.
//! These concern one K-managed identity. Safety alone does not prove liveness.
use crate::state::{Phase, Slot};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveProcess {
    pub slot: Slot,
    pub pid: u32,
    pub start_id: String,
    pub version: String,
}
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Slots {
    pub stable: Option<String>,
    pub experiment: Option<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorldSnapshot {
    pub phase: Phase,
    pub slots: Slots,
    pub live_processes: Vec<LiveProcess>,
    pub journal_intents: Vec<Phase>,
    pub workload_digest: Option<String>,
    pub prior_incarnation_start_id: Option<String>,
    pub install_ownership: Option<Ownership>,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Ownership {
    #[serde(rename = "self")]
    SelfOwned,
    ManagedElsewhere,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum HostAssumption {
    QuiesceResumeInverse,
    ProbeFromLiveProcess,
    CompatibilityDeclared,
    DataFormatBackwardCompatible,
    ExclusiveHandoff,
    ResidentService,
}
pub struct Invariant {
    pub id: &'static str,
    pub assumes: &'static [HostAssumption],
    pub check: fn(&WorldSnapshot) -> Option<String>,
}
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Violation {
    pub invariant_id: String,
    pub reason: String,
}
pub const BUILT_IN_INVARIANTS: &[Invariant] = &[
    Invariant {
        id: "k.never-dual-run",
        assumes: &[HostAssumption::ExclusiveHandoff],
        check: |s| {
            (s.live_processes.len() > 1)
                .then(|| format!("{} live incarnations", s.live_processes.len()))
        },
    },
    Invariant {
        id: "k.never-bricked",
        assumes: &[],
        check: |s| {
            (s.slots.stable.is_none() && s.slots.experiment.is_none())
                .then(|| "both slots empty".into())
        },
    },
    Invariant {
        id: "k.live-process-matches-slot",
        assumes: &[],
        check: |s| {
            s.live_processes.iter().find_map(|p| {
                let version = match p.slot {
                    Slot::Stable => &s.slots.stable,
                    Slot::Experiment => &s.slots.experiment,
                };
                version
                    .as_ref()
                    .filter(|v| **v != p.version)
                    .map(|v| format!("slot holds {v} but live process reports {}", p.version))
            })
        },
    },
    Invariant {
        id: "k.journal-precedes-phase",
        assumes: &[],
        check: |s| {
            (s.phase != Phase::Idle && !s.journal_intents.contains(&s.phase))
                .then(|| format!("phase {:?} has no journal intent", s.phase))
        },
    },
    Invariant {
        id: "k.terminal-leaves-no-experiment",
        assumes: &[],
        check: |s| {
            (matches!(s.phase, Phase::Promoted | Phase::RolledBack) && s.slots.experiment.is_some())
                .then(|| "terminal transaction retains experiment".into())
        },
    },
    Invariant {
        id: "k.managed-copy-never-self-upgrades",
        assumes: &[],
        check: |s| {
            (s.install_ownership == Some(Ownership::ManagedElsewhere) && s.phase != Phase::Idle)
                .then(|| "externally managed installation entered a transaction".into())
        },
    },
];
pub fn check_invariants(snapshot: &WorldSnapshot, invariants: &[Invariant]) -> Vec<Violation> {
    invariants
        .iter()
        .filter_map(|i| {
            (i.check)(snapshot).map(|reason| Violation {
                invariant_id: i.id.into(),
                reason,
            })
        })
        .collect()
}
pub const WORKLOAD_PRESERVED_ASSUMES: &[HostAssumption] = &[HostAssumption::QuiesceResumeInverse];
pub fn workload_preserved(before: &WorldSnapshot, after: &WorldSnapshot) -> Option<String> {
    match (&before.workload_digest, &after.workload_digest) {
        (Some(a), Some(b)) if a != b => Some("workload digest changed across transition".into()),
        _ => None,
    }
}
fn terminal(phase: Phase) -> bool {
    matches!(phase, Phase::Idle | Phase::Promoted | Phase::RolledBack)
}
pub fn reaches_terminal_within(trace: &[Phase], max_steps: usize) -> Option<String> {
    if trace.last().is_some_and(|p| terminal(*p)) {
        return None;
    }
    (trace.len() >= max_steps).then(|| {
        format!(
            "transaction has not reached terminal after {} steps (limit {max_steps})",
            trace.len()
        )
    })
}
pub const SERVICE_SETTLES_LIVE_ASSUMES: &[HostAssumption] = &[HostAssumption::ResidentService];
pub fn service_settles_live(
    observations: &[WorldSnapshot],
    max_observations: usize,
) -> Option<String> {
    let first = observations.iter().position(|s| terminal(s.phase))?;
    for s in observations.iter().skip(first).take(max_observations) {
        if s.slots
            .stable
            .as_ref()
            .is_some_and(|v| s.live_processes.iter().any(|p| &p.version == v))
        {
            return None;
        }
    }
    Some(format!(
        "terminal transaction has no live stable incarnation within {max_observations} observations"
    ))
}
